BUG - Fastlane Unit Tests #173

anupdsouza opened the pull request:
# What it Does
* Closes #173 
* Provides coverage report #71 using xcov

# How I Tested
* Executed unit tests locally using fastlane, `bundle exec fastlane unit_tests`
* Ran github workflow after pushing to branch

# Notes
* Removed `build_for_testing` config setting which was only building the code but not executing the tests. [Ref fastlane docs.](https://docs.fastlane.tools/actions/run_tests/).
* Updated gemfile to install `xcov` gem for code coverage.
* Added `xcov` action in fastfile to provide code coverage.
* Minimum code coverage has not been specified yet as there aren't enough tests written. This can be updated later by adding `minimum_coverage_percentage: 80.0` with an acceptable benchmark value to the `xcov` action in the fastfile.
* Added `.xcovignore` file in fastlane folder. This can be used to exclude files we do not need to test or require coverage to be gathered for. I believe model/business logic files are more suited for testing. This can be updated as development progresses.
* Added `xcodebuild_formatter` config in fastlane to use `xcbeautify` for cleaner logs.
* Updated github workflow to install dependencies via bundle (for gems) & via homebrew (for xcbeautify)
* Right now the plan only executes unit tests via the `only_testing` config flag in fastlane. If removed, it will also execute UI tests ( #54 ) but I've observed the workflow taking over an hour to complete on GitHub. So maybe it's something we need to weigh the pros & cons for?
* Updated workflow such that only the last initiated workflow executes in order to prevent multiple workflows from running on multiple code push actions. See `concurrency` in workflow file.
* In order to run locally, make sure to first run `bundle update` for installing gem dependencies and then run `brew install xcbeautify` to install xcbeautify for cleaner logging. You might need to use `sudo` incase you run into errors for write permissions to the ruby directory

# Screenshots ~ Local
<img width="1086" alt="Screenshot 2023-10-20 at 5 26 24 PM" src="https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/103429618/79be560c-2e44-4603-9418-9d5fec9e01ff">

<img width="849" alt="Screenshot 2023-10-20 at 5 26 48 PM" src="https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/103429618/a90a4eb6-3460-45a7-b78a-8edaae2c927b">

<img width="490" alt="Screenshot 2023-10-20 at 5 27 48 PM" src="https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/103429618/49492f87-36ff-45d4-998a-3ea0dcce9f6d">

Coverage report ~ `fastlane/CodeCoverageReport/index.html`
---
<img width="1680" alt="Screenshot 2023-10-20 at 5 28 18 PM" src="https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/103429618/16b8a596-00b1-4c84-8067-89da7520224e">


# Screenshots ~ Github workflow
---
Failing tests
---
<img width="1199" alt="Screenshot 2023-10-20 at 10 03 20 AM" src="https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/103429618/1c57d7c7-2808-4a94-82e6-874636e1f0ab">

<img width="1189" alt="Screenshot 2023-10-20 at 10 03 31 AM" src="https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/103429618/022ed6d3-4f61-4421-ba8d-26271efcac60">

Passing tests
---
<img width="1225" alt="Screenshot 2023-10-20 at 9 48 38 AM" src="https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/103429618/88d1397e-54c0-4ae5-8ede-b19737ea6ae0">

<img width="1124" alt="Screenshot 2023-10-20 at 1 06 11 PM" src="https://github.com/mikaelacaron/Basic-Car-Maintenance/assets/103429618/89a37b74-d485-4ecd-86ff-f7d8e479d9f1">

# Miscellaneous
* We can add a unit test status badge to the ReadMe as well to provide a quick status of the unit testing workflow. It'd look something like this ![Unit Tests Workflow Badge](https://github.com/mikaelacaron/Basic-Car-Maintenance/actions/workflows/unit-tests.yml/badge.svg) but it would say passing or failing depending on the unit test workflow status. If you feel this adds value, please feel free to create an enhancement or improvement & assign to me & I'd be happy to take it up!

anupdsouza committed:
Update unit test workflow configuration

Removed build_for_testing config setting
Added xcov action for code coverage & excluded view files
Update gemfile with xcov gem
Update workflow to install dependencies

anupdsouza committed:
Include only project targets for code coverage

mikaelacaron commented:
Thanks for the detailed PR!!
> Right now the plan only executes unit tests via the only_testing config flag in fastlane. If removed, it will also execute UI tests ( https://github.com/mikaelacaron/Basic-Car-Maintenance/issues/54 ) but I've observed the workflow taking over an hour to complete on GitHub. So maybe it's something we need to weigh the pros & cons for?

Yeah that's definitely an issue, let's not run them at this moment, and I may just run them manually for now (when it's needed)

> We can add a unit test status badge to the ReadMe as well to provide a quick status of the unit testing workflow. It'd look something like this Unit Tests Workflow Badge but it would say passing or failing depending on the unit test workflow status. If you feel this adds value, please feel free to create an enhancement or improvement & assign to me & I'd be happy to take it up!

Yes I'll do this after I merge your PR!

I'll look at this soon, I haven't gone through the code changes, and I need to run it all locally still

mikaelacaron commented on fastlane/Fastfile:
```diff
@@ -1,30 +1,35 @@
-#
-
-# Uncomment the line if you want fastlane to automatically update itself
-# update_fastlane
```
Please keep this here! I'm not super familiar with fastlane, and this is still useful

anupdsouza replied:
Done.

anupdsouza replied:
Done.

mikaelacaron commented on fastlane/Fastfile:
```diff
@@ -1,30 +1,35 @@
+		xcov(
+			scheme: "Basic-Car-Maintenance",
+			only_project_targets: true,
+			output_directory: "./fastlane/CodeCoverageReport"
```
Does this need to be ignored?

anupdsouza replied:
Not sure what you mean? `xcov` is the fastlane action that runs coverage for the project.

anupdsouza replied:
Updated gitignore file to exclude reports from accidentally being committed to the repo

mikaelacaron reviewed (changes requested):
> Minimum code coverage has not been specified yet as there aren't enough tests written. This can be updated later by adding minimum_coverage_percentage: 80.0 with an acceptable benchmark value to the xcov action in the fastfile.

Can you add this in the right spot, but comment it out?

I have a few questions too, noted below

I haven't been able to run this locally, `bundle update` has been timing out

anupdsouza committed:
Update fastfile

Put back default fastfile template
Add & comment out minimum_coverage_percentage flag

anupdsouza committed:
Update .gitignore

mikaelacaron reviewed (approved):
Great work!! I got it to run locally! Not sure what happened the first time

Thanks for your patience!

mikaelacaron committed:
Add fastlane's readme

mikaelacaron committed:
Add Gemfile.lock
